### PR TITLE
Feature/implement validation

### PR DESCRIPTION
**Pull request by romaninsh**
Implements a consistent validation for models.

For examples, see https://github.com/atk4/data/blob/3ab7741e8d0956671c3c309d9f55ac963faefa1f/src/ValidationException.php

TODO: documentation

### CODE REVIEW

File: src/Model.php [hunk 1]
```diff
@@ -1249,6 +1275,9 @@ public function save($data = [], Persistence $to_persistence = null)
         }
 
         return $this->atomic(function () use ($to_persistence) {
+            if (($errors = $this->validate('save')) !== []) {
+                throw new ValidationException($errors);
```
Comment (DarkSide666) [inline src/Model.php]: Probably you should add 2nd parameter here like this `throw new ValidationException($errors, 'save');`

File: src/ValidationException.php [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+            foreach ($errors as $field=>$error) {
+                return parent::__construct([
+                    $error,
+                    'field'=> $field,
+                ]);
+            }
+        }
+
+        return parent::__construct([
+            'Incorrectly use of ValidationException, argument should be an array',
```
Comment (DarkSide666) [inline src/ValidationException.php]: "Incorrect use ..." not "Incorrectly use ..." :)

File: src/Model.php (hunk 1 above)
Comment (romaninsh) [inline src/Model.php]: tried to think of a good reason why, but just cant find

File: src/ValidationException.php (hunk 2 above)
Comment (romaninsh) [inline src/ValidationException.php]: fixed

File: src/Model.php (hunk 1 above)
Comment (DarkSide666) [inline src/Model.php]: because ValidationException use and show 2nd parameter :)

**Pull request merged**